### PR TITLE
Use rustc's error reporting mechinery

### PR DESCRIPTION
**Pull request by DevinR528**
Enables using `rustc_lint::LateContext` to build warnings/errors for us from our `Lint` and `dyn Span` types.

If you want I can clean up the first few commits, so it's just, fix typos, change `Lint` as one commit, then the unsafe code, and change code in `linter_lints`?

### CODE REVIEW

**Comment by xFrednet** [comment]
The last PR created some conflics, which should be simple to fix. Could you rebase?

Currently I'm debating different ways to implement lint crate tests. When done this should enable us to test out lints from `linter_lints` and maybe have lint creates in out test folder as well :)

I'll still try to get to a review today, as the conflics are simple enough to not effect the overall implementation :)

File: linter_driver_rustc/src/ast/rustc.rs
Comment (DevinR528) [inline linter_driver_rustc/src/ast/rustc.rs:17]: I made this a single commit because I'm not sure it's better to remove the fields. `LateContext` has the `LintStore` and `TyCtxt` so it seems a bit redundant? We need the `LateContext` if we want to use the warn/error reporting of rustc.

Comment (xFrednet) [reply linter_driver_rustc/src/ast/rustc.rs:17]: I remember not using `LateContext` due to some lifetime problems, as it doesn't live for `'tcx` but this might no longer be important. Even without it we could span a lint with [`struct_span_lint_hir`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.struct_span_lint_hir). That is also the function used by `LateContext` internally. But if everything works like this, it should be fine :)

**Comment by xFrednet** [comment]
> Example output so you don't have to run it yourself
> 
> ```
> warning: hey there is a static item here
>   --> linter_api/src/lib.rs:10:1
>    |
> 10 | pub static LINTER_API_VERSION: &str = env!("CARGO_PKG_VERSION");
>    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>    |
>    = note: `#[warn(test_lint)]` on by default
> ```

Nice! I'm currently thinking of ways to implement simple UI tests for the user, but also for us. Then we can hopefully also have tests like rustc and Clippy uses :)

File: linter_driver_rustc/src/ast/rustc.rs
Comment (xFrednet) [inline linter_driver_rustc/src/ast/rustc.rs:76]: This `to_leaked_rustc_lint()` call means that every time this function is called, a new lint instance is created and leaked. This is a bit problematic IMO. Another problem is, that rustc internally uses the memory address of lints for comparison, since these are static.

Ideally, we would create these lint instances in [linter_driver_rustc/src/main.rs](https://github.com/rust-linting/rust-linting/blob/1ed2c0a96ef282bf65f568bd18fc43d3faf039ab/linter_driver_rustc/src/main.rs#L40-L45) so that rustc can track the lint level. The issue with this is, that we would then have to somehow store all created lints and later save them in `RustcContext` and use them for emission. This in turn requires the lint modules to be loaded at that point as well. I've tried to do that, but ran into lifetime problems...

As an intermediate solution, I think that we should add a `HashMap` to `RustcContext` that can map from our lint instances to Rustc's leaked lint instances. For the emission, we would then only create a new instance if it's the first lint trigger and otherwise reuse the already create one.

Does this make sense? :upside_down_face:

File: linter_driver_rustc/src/ast/rustc.rs
Comment (xFrednet) [inline linter_driver_rustc/src/ast/rustc.rs:29]: Another tiny nit: The name should ideally have some kind of prefix. Later, I would like something that indicates the lint crate and maybe that it comes from our tool. For now I think that `linter` would be a good prefix. Then the emission would say `` `#[warn(linter::test_lint)]` on by default``

**Review by xFrednet — COMMENTED** [review]
I ran it locally and like the output. I've added some comments, hopefully not too nit-picky. For the difficult ones, I've tried to also add a simple workaround solution to not hold this PR up :)

**Comment by DevinR528** [comment]
It might take a day or 2 for me to get to this (just moved down to Texas for a summer internship) but I agree with all the feedback (also not sure why I didn't fix some of the stuff it seems like was a repeat of my earlier draft PR, sorry about that, I guess copy/pasting is never as great as you think :laughing:). 

I had one thing that I noticed while working on this, how do you feel about the name for `lint_driver_rustc::ast::RustcWhatever` (replace "Whatever" with the name of an item type i.e. Item, Ty, etc)? I can't think of a better one. They are the concrete types for the traits so maybe `ConcreteWhatever` or `BackingWhatever` (neither very good) :shrug: just a thought, I initially thought they were renames of the actual rustc types `use rustc_middle::Item as RustcItem;` kinda thing.

**Comment by xFrednet** [comment]
> It might take a day or 2 for me to get to this (just moved down to Texas for a summer internship)

That's totally fine. I hope the internship will be fun and educational :).

> I guess copy/pasting is never as great as you think :laughing:).

Good old copy/pasta :spaghetti: :joy: 

> I had one thing that I noticed while working on this, how do you feel about the name for `lint_driver_rustc::ast::RustcWhatever` (replace "Whatever" with the name of an item type i.e. Item, Ty, etc)? I can't think of a better one. They are the concrete types for the traits so maybe `ConcreteWhatever` or `BackingWhatever` (neither very good) shrug just a thought, I initially thought they were renames of the actual rustc types `use rustc_middle::Item as RustcItem;` kinda thing.

For most items in `linter_api::ast::` I actually want to move away from traits. For reasons documented in #6. With that, we'll have less direct structs in `lint_driver_rustc::ast::` than we have right now. But we'll still have some, for instance for the context. My initial idea with this naming convention, was to provide us with information, which driver created an item. I see your point, that these at first give the impression of just been an alias. By now I also think that it's unimportant to store the driver, since there will only be one active at the time, and right now, we only have rustc.

Maybe we could go the Java/C# route and call them `WhateverImpl`. I've seen some instances, where this didn't fit in java, but it worked well enough. Another Option could be `DriverWhatever`, for me, it's also fine to keep `RustcWhatever`.

What are your thoughts on the matter? (We might want to move this to a separate issue, to keep track of this)

File: linter_driver_rustc/src/ast/rustc.rs
Comment (DevinR528) [reply linter_driver_rustc/src/ast/rustc.rs:15]: Perfect, I like it!

File: linter_api/src/context.rs
Comment (DevinR528) [reply linter_api/src/context.rs:34]: Like just have an easier way to let the user emit a lint (easier compared to rustc now, I feel like it was difficult to discover the way to do this using the rustic API for the first time) so I'm with you for docs and discovery a super easy aliased or just a pass-through method kinda thing would be nice.

```rust
trait DriverContext { 
    fn warn(&self, msg: &str) {
        /* Somehow access the static Lint from the user  and fill in the rest for them*/
        self.emit_lint(msg, get_static_lint_and_convert()); 
    }
}
```

Comment (DevinR528) [reply linter_api/src/context.rs:34]: ps I changed the names to `emit_lint[*]`

File: linter_driver_rustc/src/ast/rustc.rs
Comment (DevinR528) [reply linter_driver_rustc/src/ast/rustc.rs:29]: I think the way clippy does this is like
`name: concat!("linter::", stringify!($NAME)),`
for the line in the macro right?

File: linter_api/src/context.rs
Comment (xFrednet) [reply linter_api/src/context.rs:34]: I feel like documentation and lint development tools, will be a major part of this project. I'm trying to already provide some helpful doc comments, but it's hard to know how everything will look like in the end. Thank you for the rename, let's keep this in mind for later :+1:

File: linter_driver_rustc/src/ast/rustc.rs
Comment (xFrednet) [reply linter_driver_rustc/src/ast/rustc.rs:29]: Yes :+1:, rustc also does that in the tool lint macro :upside_down_face:

File: linter_driver_rustc/src/ast/rustc.rs
Comment (DevinR528) [reply linter_driver_rustc/src/ast/rustc.rs]: Well I had to also wrap the FxHashMap in a RefCell since we call it from a method that is `&self` which I think we should keep it that way. Eventually maybe this will be `Arc<Mutex<...>>` and we can run the lints in parallel and just merge the outputs nicely somehow keeping track of ordering :shrug:

anyways I'm pretty sure this is what you meant, good catch this would have been a LOT of leaking oops haha.

File: linter_driver_rustc/src/ast/rustc.rs
Comment (xFrednet) [reply linter_driver_rustc/src/ast/rustc.rs:83]: I have an idea for this specific case, which I'll work on, when I remove the span trait in favor of a struct (See #6). For now, this is as said fine

**Comment by xFrednet** [comment]
Thank you for addressing the feedback. I'll do a full review tomorrow. If you want, you can rebase on master and add a UI test :upside_down_face:

File: linter_driver_rustc/src/ast/rustc.rs
Comment (xFrednet) [inline linter_driver_rustc/src/ast/rustc.rs:61]: Small note about the function signature: we should think about users create diagnostics. Clippy has several util functions like `span_lint`, `span_lint_and_note` etc. This forces the user to always create the suggestion before the emission, even it the lint is allowed. This can also cause some ICEs, which is why Clippy wants to move away from this model. (There is an issue for this, even if there has been no progress AFAIK).

I like the diagnostic build approach of rustc. For now, this is a good starting point. I'll be busy over the next two weeks, but this is one thing I want to discuss or at least create an issue for in the design repo. :upside_down_face:

**Review by xFrednet — APPROVED** [review]
This looks good to me, I've added one informative comment. Could you rebase on master and add a UI test? Then everything should be ready. Thank you for the help :+1:

File: linter_driver_rustc/src/ast/rustc.rs
Comment (DevinR528) [reply linter_driver_rustc/src/ast/rustc.rs:61]: I'm not sure I understand what could be the other way of doing it. Or maybe I'm misunderstanding what you mean by suggestion and emission? An issue sounds like a good idea anyway better place to continue the discussion!

**Comment by DevinR528** [comment]
Shuld be ready to go, you want me to merge this or should I wait for you in the future (once you have approved it should I feel ok to merge?)

**Review by xFrednet — APPROVED** [review]
Ähm, good question. Usually, if all that needs to be done is a rebase, it's okay to `r=me`. In those cases, I try to also mention that. In this cases, I wanted to see the added uitest, also to make sure that everything works as intended.

Now everything looks good, and you can r=me after another rebase. The number of conflicts should hopefully decrease with the growth of the project :upside_down_face:

**Comment by DevinR528** [comment]
bors r=xFrednet

**Pull request merged**
